### ISSUE TITLE
Permit use of user-generated .db3 databases/utilizing the PIF json import

### ISSUE THREAD

**Issue by Mostelin** [christofmuc/KnobKraft-orm#121]
I have had a look at the structure of your Knobkraft ORM database files and would like to have a go at generating one directly from Python. This is because I have about 6000 programs (in an MS-Access database) which I process with Python. This set-up doesn't provide a GUI though, nor does it work with MIDI, while the ORM does both. Ideally, I could get my vast database of programs into the ORM without having to do all the importing manually.

I wonder whether the IDs you create for your imports can be generated in Python, and would they work?

I also wonder whether it would be possible to generate an alternative automatic_categories.jsonc file so that I could preserve or reproduce the categorisation already built into my database?

While I play with Python, I am not a programmer and am hesitant to get immersed in compiling languages.

**Comment by christofmuc** [comment]
Good idea! And multiple questions here, let me try to sort:

1. Directly creating a SQLite database certainly is possible, and there is no reason you couldn't open it with the Orm if it is error free. It is just probably not very stable, as the database could change and there might be surprises waiting with platform dependence or I don't know what.

It could be much simpler: you could generate a so called "Patch Interchange File", which is just a JSON file in a syntax I invented that can be bulk imported into the Orm, and will nicely populate the database for you. It is read in by the same "import file from computer" as is used for standard Sysex files, just with a json extension.

The format is pretty simple JSON, and I don't recall documenting it properly yet, but you could export a few of your existing patches into the PIF format with that button in the software, then you see how it looks like. The idea for this is to have a full roundtrip available from SQL database into PIF and back again without losing any data.

So you could write a little Python script opening the Access database and writing out a JSON file with all of them. This should be pretty straightforward, let me know if you need help getting it started (or finished :-)

2. Don't worry about the IDs - those are strings that should be unique. The Orm contains a mechanism to reindex (recreate the IDs) in case the ID algorithm changes, I would assume that can be used to recalculate IDs after inital import. It basically is an MD5 sum of the relevant part (i.e. not the header, no the name, not the program position) of the data.

3. The automatic_category.json is actually meant to be editable (cautiously), and there is another file for mapping imported categories into database categories, this is already used for Synth which have the category as part of the patch like the Virus. The PIF file also contains categories, so that should also be possible. You can use the UI to add more categories to the database before you run the big import on it.

So in summary, all of this will be possible without reinventing the wheel, but to piece everything together will be a bit of work, certainly worthwhile for a collection of 6000 pathches!

**Comment by Mostelin** [comment]
Thanks for the full answer. I will start to work on PIF-file generation while I finish refining my categories. I looked at the PIF structure and I am curious as to how you are encoding the programs there. I haven't been able to figure that out.

**Comment by Mostelin** [comment]
The PIF import is now working - rather well, too. I've imported 5,762 programs and the program handles the load. Moreover, the synth accepts the programs and plays them perfectly. It's very nice to be able to open the left panel to select by original bank file and to see how many of each bank's programs survived the de-duplication cull. And showing (on the right) the original bank position of each program helps if you want to refer back to original copies of programs.

Stuff I am still working on:
- Accomodating my categories - most of them are being ignored and I don't know why.
- Keeping SQ-80 programs apart from those of the ESQ-1. It looks like I must create a separate adaptation if I want this.
- Trivial, but puzzling: removing the comma after the last program in the JSON file. I don't know how to anticipate EOF.

**Comment by christofmuc** [comment]
@Mostelin Let me think 

- the categories have to pre-exist in the database as I recall, the PIF import will not create new categories on the fly. You could use the UI to create your categories before running the import. There is also the option to create a mapping file to map a synth's category into the existing one. I think I did that for the Access Virus, maybe have a look into the file mapping_categories.jsonc, this is what is opened in an editor when you use the Categories... edit category import mapping menu item. Not sure if this works, it does for the Virus for sure.
- Yes, to have SQ-80 programs not to be mixed with ESQ-1 you would need a new adaptation, as the synth name is the primary key. Are they fully compatible or just one way? For fully compatible, I always only setup one adaptation (like for the Korg DW8000/EX8000)
- ha! the simple things. wouldn't be a problem if you don't "print" the json yourself, but rather create a json structure in memory and then serialize that to a string, in python that's dumps() iirc.

**Comment by Mostelin** [comment]
I must admit to struggling with the categories of the ORM. I think I have three separate problems:

1.  When I add categories (menu: 'Edit auto-categories'), at some point the new entries start to appear in two place: at the bottom of the list, and overwriting entries at the top of the list too. I don't think this is reversible.
2. When I edit mapping_categories.jsonc, the new mappings are ignored. See example attached, where I have changed the synthesizer name to 'Ensoniq ESQ-1' (though this seems to make no difference), and added a line to map 'Woodwind' to 'Wind'. [mapping_categories.zip](https://github.com/christofmuc/KnobKraft-orm/files/7655106/mapping_categories.zip)
3.  When I look for a local copy of automatic_category.json to troubleshoot, I come up empty-handed. Is it compiled in?

**Comment by christofmuc** [comment]
@Mostelin Hum, let me try to dissect:

1. Thanks for pointing this out! This seems to be a regression bug, in your version you can actually only add one category and then have to press Ok to save it, and reopen the dialog. Will be fixed with the next release so you can again add multiple categories in one go. Not sure when this crept in or if it never worked :-/
2. The new mappings probably can't have worked if the edit category dialog was so broken. I'd say we try again with the new version I'll do today. BTW, to debug the database I use the DBBrowser for SQLite, which is awesome to quickly look at what is really stored in the database.
3. The automatic_category.jsonc has nothing to do with the category mapping, just with automatic detection of categories based on regular expressions matching patch names :-) The file is embedded (compiled in), but can be overridden placing it in the appdata directory (you'll see a log message if it is overridden). The logic to automatically create the override file is currently disabled as I wanted to expand the edit category dialog to add editing the regular expressions there, but never got around to do it. But as said, this is only the file for name parsing.

**Comment by Mostelin** [comment]
1. I will have another go at this in the morning. You may well have it fixed by then, but if not I'll do things one at a time.
2. Let's see how it goes with the new attempt.
3. OK, that confirms my guesses. I was just looking for another potential category-related knob to play with. It's good to know that placing a copy in the appdata directory creates an override. I haven't got to the bottom of it yet, but I noticed some patterns in naming emerging as I worked through my long list of programs, over and over again. I may be able to extend the regular expressions, but if not it won't really matter.

**Comment by christofmuc** [comment]
@Mostelin While preparing the new fixed release, I read through this issue and decided all of the questions have been answered or lead otherwise to bug reports in the other ticket, so I close this. Feel free to amend, of course, as usual feedback is always appreciated!

**Comment by Mostelin** [comment]
Yes, I agree that this is all either resolved or covered in other (categories) tickets now.
